src/v/cluster/rm_stm.cc

```diff
@@ -1009,47 +1012,184 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate_seq(
   model::batch_identity bid,
   model::record_batch_reader br,
   raft::replicate_options opts,
-  [[maybe_unused]] ss::lw_shared_ptr<available_promise<>> enqueued) {
+  ss::lw_shared_ptr<available_promise<>> enqueued) {
     if (!co_await sync(_sync_timeout)) {
         co_return errc::not_leader;
     }
     auto synced_term = _insync_term;
-
+    
     if (bid.first_seq > bid.last_seq) {
-        vlog(
-          clusterlog.warn,
-          "first_seq={} of the batch should be less or equal to last_seq={}",
-          bid.first_seq,
-          bid.last_seq);
-        outcome->set_value();
+        vlog(clusterlog.warn, "first_seq={} of the batch should be less or equal to last_seq={}", bid.first_seq, bid.last_seq);
         co_return errc::generic_tx_error;
     }
 
+    // getting session by client's pid
+    ss::lw_shared_ptr<inflight_requests> session;
+    auto session_it = _inflight_requests.find(bid.pid);
+    if (session_it == _inflight_requests.end()) {
+        session = ss::make_lw_shared<inflight_requests>();
+        _inflight_requests.try_emplace(bid.pid, session);
+    } else {
+        session = session_it->second;
+    }
+
+    // critical section, rm_stm evaluates the request its order and
+    // whether or not it's duplicated
+    // ASSUMING THAT ALL INFLIGHT REPLICATION REQUESTS WILL SUCCEED
+    // if any of the request fail we expect that the leader step down
+    auto u = co_await session->lock.get_units();
+
+    if (!_c->is_leader() || _c->term() != synced_term) {
+        u.return_all();
+        co_return errc::not_leader;
+    }
+
+    if (session->term < synced_term) {
+        // we don't care about old inflight requests because the sync
+        // guarantee (all replicated records of the last term are
+        // applied) makes sure that the passed inflight records got
+        // into _log_state->seq_table
+        session->forget();
+        session->term = synced_term;
+    }
+    
+    // checking if the request (identified by seq) is already resolved
+    // checking among the pending requests
+    auto cached_r = session->known_seq(bid.last_seq);
+    if (cached_r) {
+        u.return_all();
```

Comment:
neither of those `u.return_all()` are not required as units are leaving scope

Reply:
nice catch, I've removed it in all places where it's redundant